internal/backend/server.go

```diff
@@ -847,21 +849,123 @@ func setRealIPRequest(req *http.Request) *http.Request {
 }
 
 type server struct {
-	server *http.Server
-	certData
+	server   *http.Server
+	certData *certData
 }
 
 type certData struct {
+	cert     tls.Certificate
 	certFile string
 	keyFile  string
+	mu       sync.Mutex
+	loaded   bool
 }
 
-func (s *server) ListenAndServe() error {
-	if s.certFile != "" || s.keyFile != "" {
-		return s.server.ListenAndServeTLS(s.certFile, s.keyFile)
+func (c *certData) load() error {
+	cert, err := tls.LoadX509KeyPair(c.certFile, c.keyFile)
+
+	if err == nil {
+		c.loaded = true
+	}
+
+	if err != nil {
+		return err
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cert = cert
+
+	return nil
+}
+
+func (c *certData) getCert() (*tls.Certificate, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.loaded {
+		return nil, fmt.Errorf("cert data hasn't loaded the cert yet")
+	}
+
+	return &c.cert, nil
+}
+
+func (c *certData) runWatcher(ctx context.Context, logger *zap.Logger) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %w", err)
+	}
+	defer w.Close() //nolint:errcheck
+
+	if err = w.Add(c.certFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %w", c.certFile, err)
+	}
+
+	if err = w.Add(c.keyFile); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %w", c.keyFile, err)
 	}
 
-	return s.server.ListenAndServe()
+	for {
+		select {
+		case e := <-w.Events:
+			if !e.Has(fsnotify.Write) && !e.Has(fsnotify.Create) {
+				break
+			}
+
+			if err = c.load(); err != nil {
+				logger.Error("failed to load certs", zap.Error(err))
```

Comment:
should we keep watching here? as it might be incomplete file that we see right now (e.g. file being overwritten), but it will be "good" soon?